src/java.base/share/classes/java/io/FileDescriptor.java

```diff
@@ -67,14 +67,15 @@ public void beforeCheckpoint(Context<? extends jdk.crac.Resource> context) throw
                 ClaimedFDs claimedFDs = Core.getClaimedFDs();
                 FileDescriptor self = FileDescriptor.this;
 
-                claimedFDs.claimFd(self, self, () ->  {
-                    if (self == in || self == out || self == err) {
-                        return null;
-                    }
-                    return new CheckpointOpenResourceException(
-                        FileDescriptor.class.getSimpleName() + " " + fd + ": " + nativeDescription0(),
-                        getStackTraceHolder());
-                });
+                // Normally the claiming should be overridden by FileInputStream/FileOutputStream
+                // but in case these are collected we handle FDs 0..2 here as well.
+                if (self == in || self == out || self == err) {
+                    claimedFDs.claimFd(self, self, null);
+                }
+
```

Comment:
How would you expect to specify *which* FD should not throw, without fetching the details through native code? Would a numeric identifier (or full wildcard) be sufficient?

Reply:
Alternatively, we could change the claiming process to create a FD -> higher resource reference when the FD is accepted in the resource (constructor or setter), preventing this altogether from happening. Regrettably one FD would have to be able to reference multiple resources this way, though.

Reply:
I'd propose a numeric id, and ideally a regex for the native description, without trying to interpret that on the FD level.

Another use-case for these FD policies is to work around a forgotten claim from a higher-level Java object, which should be considered as a JDK bug. It will be uncommon and will be discouraged, but anyway, worth to have as a workaround.